### PR TITLE
Added Label Editor to Sample Detail Page

### PR DESCRIPTION
**Pull request by ColeVoelpel**
Let me know if there is any UI changes you want. I left the create label section open by default on the modal to fill it out but that is an easy change if necessary.

### CODE REVIEW

File: client/src/js/samples/components/LabelEditor/ColorSelector.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+import styled from "styled-components";
+import { Button, Input, InputGroup, InputLabel, InputError } from "../../../base";
+
+export const StyledColorButton = styled(Button)`
+    background-color: ${props => props.color};
+    margin: 0px 2px;
+`;
+
+export const StyledInput = styled(Input)`
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/ColorSelector.js:10]: `ColorSelectorInput`

File: client/src/js/samples/components/LabelEditor/ColorSelector.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import React from "react";
+import styled from "styled-components";
+import { Button, Input, InputGroup, InputLabel, InputError } from "../../../base";
+
+export const StyledColorButton = styled(Button)`
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/ColorSelector.js:5]: `ColorSelectorButton`

File: client/src/js/samples/components/LabelEditor/ColorSelector.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+`;
+
+const getInitialState = ({ color }) => ({
+    color: color || ""
+});
+
+export class ColorSelector extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = getInitialState(this.props);
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/ColorSelector.js:21]: You are managing state here and in your parent component.

Keep the state in `CreateLabel`. The `ColorSelector` component should take the color value as a prop and pass events back to `CreateLabel` using your `onColorChange` prop.

File: client/src/js/samples/components/LabelEditor/LabelEditor.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+import { connect } from "react-redux";
+import styled from "styled-components";
+import { get, map } from "lodash-es";
+import { Table, ViewHeader, ViewHeaderTitle, RemoveModal, Button, BoxGroup, BoxGroupHeader, Flex } from "../../../base";
+import { getLabels, createLabel, removeLabel, updateLabel } from "../../actions";
+import { CreateLabel } from "./CreateLabel";
+import { LabelItem } from "./LabelItem";
+import { EditLabel } from "./EditLabel";
+
+export const StyledTable = styled(Table)`
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelEditor.js:11]: Only use `table` elements for actual tables. You can solve this layout problem better with [flex box](https://css-tricks.com/snippets/css/a-guide-to-flexbox/).

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    max-width: 75px;
+`;
+
+export class LabelItem extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = getInitialState(this.props);
+    }
+
+    onRemove = (id, name) => {
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:50]: `handleRemove()`

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        this.props.removeLabel(id, name);
+    };
+
+    onEdit = (id, name, color, description) => {
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:54]: `handleEdit()`

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        this.props.editLabel(id, name, color, description);
+    };
+
+    render() {
+        const name = this.state.name;
+        const color = this.state.color;
+        const description = this.state.description;
+        const id = this.state.id;
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:62]: Use object spread.

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                        tip="Edit"
+                    />
+                    <StyledIcon color="red" onClick={() => this.onRemove(id, name)} name="fas fa-times" tip="Remove" />
+                </IconColumn>
+            </tr>
+        );
+    }
+}
+const mapStateToProps = state => ({
+    name: sate.name,
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:83]: Not `sate`.

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+};
+
+export const StyledLabel = styled(Label)`
+    padding: 5px;
+    font-size: ${props => props.theme.fontSize.lg};
+    background-color: ${props => props.color};
+    color: ${getContrastColor};
+`;
+
+export const StyledIcon = styled(Icon)`
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:28]: `LabelItemIcon`

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+export const Description = styled.p`
+    padding: 0px 8px;
+    margin: 0px;
+    display: inline-block;
+`;
+export const IconColumn = styled.td`
+    text-align: right;
+    min-width: 50px;
+    max-width: 75px;
+`;
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:42]: I don't think these need to be exported.

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    const yiq = (red * 299 + green * 587 + blue * 114) / 1000;
+    return yiq >= 128 ? "black" : "white";
+};
+
+export const StyledLabel = styled(Label)`
+    padding: 5px;
+    font-size: ${props => props.theme.fontSize.lg};
+    background-color: ${props => props.color};
+    color: ${getContrastColor};
+`;
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:26]: Does not need to be exported.

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+export const StyledIcon = styled(Icon)`
+    padding: 0px 8px;
+    font-size: 20px;
+`;
+
+export const Description = styled.p`
+    padding: 0px 8px;
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:34]: When CSS value is `0`, omit units.

```css
padding: 0 8px;
```

File: client/src/js/samples/components/LabelEditor/LabelItem.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+export const getContrastColor = props => {
+    const red = parseInt(props.color.substr(1, 2), 16);
+    const green = parseInt(props.color.substr(3, 2), 16);
+    const blue = parseInt(props.color.substr(5, 2), 16);
+    const yiq = (red * 299 + green * 587 + blue * 114) / 1000;
+    return yiq >= 128 ? "black" : "white";
+};
+
+export const StyledLabel = styled(Label)`
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/LabelItem.js:21]: `Styled` prefix for components is only used when it is the top-level component in another component. This should be renamed. Something like `LabelItemExample`.

**Review by igboyes — CHANGES_REQUESTED** [review]

File: client/src/js/samples/components/LabelEditor/CreateLabel.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
+import React from "react";
+import { connect } from "react-redux";
+import { InputGroup, InputLabel, Input, Button, InputError, Modal, ModalHeader, ModalBody } from "../../../base";
+import { ColorSelector } from "./ColorSelector";
+
+const getInitialState = ({ labelName, color, description, errorName, errorColor }) => ({
```
Comment (igboyes) [inline client/src/js/samples/components/LabelEditor/CreateLabel.js:6]: Do not need to derive state from props. Form will be empty initially.

`constructor` method is not needed for class.

**Comment by ColeVoelpel** [comment]
The only change I could not do was the "createLables" -> "create" name change. It is sharing the same api file with the samples which has a "create" already for creating samples.

**Review by igboyes — CHANGES_REQUESTED** [review]
Please move all label code to new top-level module at `src/js/labels`.

If you have no children for a component call it without a closing tag:
```jsx
<div>
  <Input name="test" value="Foo" />
</div>
```

Instead of:
```jsx
<div>
  <Input name="test" value="Foo"></Input>
</div>
```

**Comment by igboyes** [comment]
Replaced by #2074

**Pull request closed without merging**